presto-hive/src/main/java/com/facebook/presto/hive/S3SelectRecordCursor.java

```diff
@@ -0,0 +1,227 @@
+import static org.apache.hadoop.hive.serde.serdeConstants.LIST_COLUMN_TYPES;
+import static org.apache.hadoop.hive.serde.serdeConstants.SERIALIZATION_DDL;
+
+class S3SelectRecordCursor<K, V extends Writable>
+        extends GenericHiveRecordCursor
+{
+    private static final String THRIFT_STRUCT = "struct";
+    private static final String START_STRUCT = "{";
+    private static final String END_STRUCT = "}";
+    private static final String FIELD_SEPARATOR = ",";
+    private static final String EMPTY_STRING = "";
+    private static final String SPACE = " ";
+
+    S3SelectRecordCursor(
+            Configuration configuration,
+            Path path,
+            RecordReader recordReader,
+            long totalBytes,
+            Properties splitSchema,
+            List<HiveColumnHandle> columns,
+            DateTimeZone hiveStorageTimeZone,
+            TypeManager typeManager)
+    {
+        super(configuration, path, recordReader, totalBytes, updateSplitSchema(splitSchema, columns), columns, hiveStorageTimeZone, typeManager);
+    }
+
+    // since s3select only returns the required column, not the whole columns
+    // we need to update the split schema to include only the required columns
+    // otherwise, Serde could not deserialize output from s3select to row data correctly
+    static Properties updateSplitSchema(Properties splitSchema, List<HiveColumnHandle> columns)
+    {
+        requireNonNull(splitSchema, "splitSchema is null");
+        requireNonNull(columns, "columns is null");
+        // clone split properties for update so as not to affect the original one,
+        // which could be used by other logic
+        Properties updatedSchema = new Properties();
+        updatedSchema.putAll(splitSchema);
+        updatedSchema.setProperty(LIST_COLUMNS, buildColumns(columns));
+        updatedSchema.setProperty(LIST_COLUMN_TYPES, buildColumnTypes(columns));
+        ThriftDDL thriftDDL = parseThriftDDL(splitSchema.getProperty(SERIALIZATION_DDL));
+        thriftDDL = filterThriftDDL(thriftDDL, columns);
+        updatedSchema.setProperty(SERIALIZATION_DDL, thriftDDLToString(thriftDDL));
+        return updatedSchema;
+    }
+
+    private static String buildColumns(List<HiveColumnHandle> columns)
+    {
+        if (columns == null || columns.isEmpty()) {
+            return EMPTY_STRING;
+        }
+        return columns.stream()
+                .map(HiveColumnHandle::getName)
+                .collect(Collectors.joining(","));
+    }
+
+    private static String buildColumnTypes(List<HiveColumnHandle> columns)
+    {
+        if (columns == null || columns.isEmpty()) {
+            return EMPTY_STRING;
+        }
+        return columns
+                .stream()
+                .map(column -> column.getHiveType().getTypeInfo().getTypeName())
+                .collect(Collectors.joining(","));
+    }
+
+    private static ThriftDDL parseThriftDDL(String ddl)
```

Comment:
nit
```suggestion
    private static ThriftDDL parseThriftDdl(String ddl)
```

(according to code-style)

Reply:
Refactored